File: API/1501-1600/1540_Volume_Supported_Linear_Regression_Trend_Modified/CS/VolumeSupportedLinearRegressionTrendModifiedStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Strategy enters long when RSI is above entry level and exits when RSI exceeds exit level.
/// </summary>
public class VolumeSupportedLinearRegressionTrendModifiedStrategy : Strategy
{
	private readonly StrategyParam<int> _rsiPeriod;
	private readonly StrategyParam<decimal> _entryLevel;
	private readonly StrategyParam<decimal> _exitLevel;
	private readonly StrategyParam<int> _signalCooldownBars;
	private readonly StrategyParam<DataType> _candleType;

	private RelativeStrengthIndex _rsi = null!;
	private decimal? _prevRsi;
	private int _cooldownRemaining;

	/// <summary>
	/// RSI period.
	/// </summary>
	public int RsiPeriod
	{
		get => _rsiPeriod.Value;
		set => _rsiPeriod.Value = value;
	}

	/// <summary>
	/// Entry threshold.
	/// </summary>
	public decimal EntryLevel
	{
		get => _entryLevel.Value;
		set => _entryLevel.Value = value;
	}

	/// <summary>
	/// Exit threshold.
	/// </summary>
	public decimal ExitLevel
	{
		get => _exitLevel.Value;
		set => _exitLevel.Value = value;
	}

	/// <summary>
	/// Minimum number of closed candles between new entries.
	/// </summary>
	public int SignalCooldownBars
	{
		get => _signalCooldownBars.Value;
		set => _signalCooldownBars.Value = value;
	}

	/// <summary>
	/// Candle type to process.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initializes a new instance of <see cref="VolumeSupportedLinearRegressionTrendModifiedStrategy"/>.
	/// </summary>
	public VolumeSupportedLinearRegressionTrendModifiedStrategy()
	{
		_rsiPeriod = Param(nameof(RsiPeriod), 14)
			.SetGreaterThanZero()
			.SetDisplay("RSI Period", "RSI calculation period", "General")
			;

		_entryLevel = Param(nameof(EntryLevel), 60m)
			.SetDisplay("Entry Level", "RSI value to enter long", "General")
			;

		_exitLevel = Param(nameof(ExitLevel), 45m)
			.SetDisplay("Exit Level", "RSI value to close long", "General")
			;

		_signalCooldownBars = Param(nameof(SignalCooldownBars), 8)
			.SetNotNegative()
			.SetDisplay("Signal Cooldown Bars", "Closed candles to wait before re-entering", "General")
			;

		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_prevRsi = null;
		_cooldownRemaining = 0;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		_rsi = new RelativeStrengthIndex { Length = RsiPeriod };
		_prevRsi = null;
		_cooldownRemaining = 0;

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(_rsi, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, _rsi);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal rsi)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (_cooldownRemaining > 0)
			_cooldownRemaining--;

		if (_prevRsi is null)
		{
			_prevRsi = rsi;
			return;
		}

		if (!IsFormedAndOnlineAndAllowTrading())
		{
			_prevRsi = rsi;
			return;
		}

		var previousRsi = _prevRsi.Value;
		var longEntry = previousRsi <= EntryLevel && rsi > EntryLevel;
		var longExit = previousRsi >= ExitLevel && rsi < ExitLevel;

		if (longExit && Position > 0)
		{
			SellMarket(Position);
			_cooldownRemaining = SignalCooldownBars;
		}
		else if (_cooldownRemaining == 0 && longEntry && Position <= 0)
		{
			var volume = Volume + (Position < 0 ? -Position : 0m);
			BuyMarket(volume);
			_cooldownRemaining = SignalCooldownBars;
		}

		_prevRsi = rsi;
	}
}

File: API/1501-1600/1540_Volume_Supported_Linear_Regression_Trend_Modified/PY/volume_supported_linear_regression_trend_modified_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy


class volume_supported_linear_regression_trend_modified_strategy(Strategy):
    def __init__(self):
        super(volume_supported_linear_regression_trend_modified_strategy, self).__init__()
        self._rsi_period = self.Param("RsiPeriod", 14) \
            .SetDisplay("RSI Period", "RSI calculation period", "General")
        self._entry_level = self.Param("EntryLevel", 60.0) \
            .SetDisplay("Entry Level", "RSI value to enter long", "General")
        self._exit_level = self.Param("ExitLevel", 45.0) \
            .SetDisplay("Exit Level", "RSI value to close long", "General")
        self._signal_cooldown_bars = self.Param("SignalCooldownBars", 8) \
            .SetDisplay("Signal Cooldown Bars", "Closed candles to wait before re-entering", "General")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(1))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._prev_rsi = None
        self._cooldown_remaining = 0

    @property
    def rsi_period(self):
        return self._rsi_period.Value

    @property
    def entry_level(self):
        return self._entry_level.Value

    @property
    def exit_level(self):
        return self._exit_level.Value

    @property
    def signal_cooldown_bars(self):
        return self._signal_cooldown_bars.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(volume_supported_linear_regression_trend_modified_strategy, self).OnReseted()
        self._prev_rsi = None
        self._cooldown_remaining = 0

    def OnStarted2(self, time):
        super(volume_supported_linear_regression_trend_modified_strategy, self).OnStarted2(time)
        rsi = RelativeStrengthIndex()
        rsi.Length = self.rsi_period
        self._prev_rsi = None
        self._cooldown_remaining = 0
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(rsi, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, rsi)
            self.DrawOwnTrades(area)

    def on_process(self, candle, rsi):
        if candle.State != CandleStates.Finished:
            return
        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1
        if self._prev_rsi is None:
            self._prev_rsi = float(rsi)
            return
        previous_rsi = self._prev_rsi
        rsi_val = float(rsi)
        long_entry = previous_rsi <= self.entry_level and rsi_val > self.entry_level
        long_exit = previous_rsi >= self.exit_level and rsi_val < self.exit_level
        if long_exit and self.Position > 0:
            self.SellMarket()
            self._cooldown_remaining = self.signal_cooldown_bars
        elif self._cooldown_remaining == 0 and long_entry and self.Position <= 0:
            self.BuyMarket()
            self._cooldown_remaining = self.signal_cooldown_bars
        self._prev_rsi = rsi_val

    def CreateClone(self):
        return volume_supported_linear_regression_trend_modified_strategy()
